abr_control/controllers/path_planners/path_planner.py

```diff
@@ -264,33 +231,91 @@ def generate_path(
         # distance is greater than our ramping up and down distance, add a linear
         # velocity between the ramps to converge to the correct position
         self.remaining_dist = None
-        if curve_length >= self.starting_dist + self.ending_dist:
-            # calculate the remaining steps where we will be at constant max_v
-            remaining_dist = curve_length - (self.ending_dist + self.starting_dist)
-            constant_speed_steps = int(remaining_dist / self.max_velocity / self.dt)
 
-            self.stacked_vel_profile = np.hstack(
-                (
-                    self.starting_vel_profile,
-                    np.ones(constant_speed_steps) * self.max_velocity,
-                    self.ending_vel_profile,
+        # if the max_v is not attainable in the distance from start to target
+        # start dropping it to reach the maximum possible velocity. This is
+        # better than scaling the profile as before, because the prior method
+        # results in a constant number of steps regarless of distance, which
+        # can really slow down short reaches.
+        searching_for_valid_velocity_profile = True
+
+        max_v = self.max_velocity
+        while searching_for_valid_velocity_profile:
+            if max_v <= 0:
+                raise ValueError
+
+            if self.starting_dist != 0:
+                # Regenerate our velocity curves if start or end v have changed
+                # if (
+                #     self.starting_vel_profile is None
+                #     or self.start_velocity != start_velocity
+                # ):
+                self.starting_vel_profile = self.vel_profile.generate(
+                    start_velocity=start_velocity, target_velocity=max_v
                 )
-            )
-            if plot:
-                self.remaining_dist = remaining_dist
-                self.dist = dist
-        else:
-            # scale our profile
-            # TODO to do this properly we should evaluate the integral to get
-            # the t where the sum of the profile is half our travel distance.
-            # This way we maintain the same acceleration profile instead of
-            # maintaining the same number of steps and accelerating more slowly,
-            # which is what we do by scaling the vel profile down
-            scale = curve_length / (self.starting_dist + self.ending_dist)
-            self.stacked_vel_profile = np.hstack(
-                (scale * self.starting_vel_profile, scale * self.ending_vel_profile)
-            )
 
+                # calculate the distance covered ramping from start_velocity to
+                # max_v and from max_v to target_velocity
+                self.starting_dist = np.sum(self.starting_vel_profile * self.dt)
+
+            if self.ending_dist != 0:
+                # if our start and end v are the same, just mirror the curve to
```

Comment:
is this code still relevant? if not should be removed

Reply:
Added back the relevant line that mirrors the ramp up when generating the ramp down if the target_v == start_v to prevent generating the same profile twice.